agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+ *
+ */
+static int nr_monolog_fmt_context_item(zval* value,
+                                       char** strbuf,
+                                       zend_hash_key* hash_key TSRMLS_DC) {
+  NR_UNUSED_TSRMLS;
+  char* key = nr_monolog_fmt_context_key(hash_key);
+  char* val = nr_monolog_fmt_context_value(value);
+
+  char* kv_str = nr_formatf("%s => %s", key, val);
+  nr_free(key);
+  nr_free(val);
+
+  char* sep = nr_strlen(*strbuf) > 1 ? ", " : "";
+  *strbuf = nr_str_append(*strbuf, kv_str, sep);
+  nr_free(kv_str);
+
+  return ZEND_HASH_APPLY_KEEP;
+}
+
+/*
+ * Purpose : Iterate over $context array and format each element
+ *
+ * Params  : string buffer to store formatted output and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_fmt_context(char* strbuf,
+                                    HashTable* context TSRMLS_DC) {
+  strbuf = nr_str_append(strbuf, "[", "");
+
+  nr_php_zend_hash_zval_apply(context,
+                              (nr_php_zval_apply_t)nr_monolog_fmt_context_item,
+                              (void*)&strbuf TSRMLS_CC);
+
+  return nr_str_append(strbuf, "]", "");
+}
+
+/*
+ * Purpose : Convert $context argument of Monolog\Logger::addRecord to a string
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_get_context(size_t argc, NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* context = nr_strdup("");
+  zval* context_arg = NULL;
+
+  if (3 > argc) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context not available", __func__);
+    goto return_context;
+  }
+
+  context_arg = nr_php_arg_get(3, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+
+  if (!nr_php_is_zval_valid_array(context_arg)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected $context be a valid array, got type %d",
+                     __func__, Z_TYPE_P(context_arg));
+    goto return_context;
+  }
+
+  if (0 == nr_php_zend_hash_num_elements(Z_ARRVAL_P(context_arg))) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context has no elements", __func__);
+    goto return_context;
+  }
+
+  context = nr_monolog_fmt_context(context, Z_ARRVAL_P(context_arg) TSRMLS_CC);
+
+return_context:
+  nr_php_arg_release(&context_arg);
+  return context;
+}
+
+/*
+ * Purpose : Combine $message and $context arguments of
+ * Monolog\Logger::addRecord into a single string to be used as a message
+ * property of the log event.
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with a log record message; caller must free
+ */
+static char* nr_monolog_build_message(size_t argc, NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* message_and_context = nr_strdup("");
+
+  char* message = nr_monolog_get_message(NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
```

Comment:
What does the spec say to do if the message is NULL or empty?

Reply:
"A non-null, non-empty message. If the message is nil, empty, or missing, or simply whitespace, the log record MUST NOT be reported."

Reply:
This spec requirement will be handled at the time when the log event is added to the transaction.